lec07-pop-models.Rmd

```diff
@@ -1,4 +1,617 @@
+
+   - Plot $\frac{dx}{dt}$ vs. $x$ for $x$ ranging from $-2$ to $2$ in $0.1$ increments.
+   - Sketch a phase portrait. Are the fixed point(s) stable or unstable?
+
+$$\frac{dx}{dt} = x^2 - 1$$
+<!--
+```{r}
+library(ggplot2)
+
+x_seq = seq(-2, 2, by = 0.1)
+
+dx_dt = x_seq^2 - 1
+
+qplot(x_seq, dx_dt) +
+  geom_line(aes(x = x_seq, y = 0))
+```
+-->
+
+
+## Numerically simulating models in R
+
+I'm moving this topic ahead so that we can use what we learn here to aid us in analyzing more
+complicated models in one and two dimensions. 
+
+Some differential equations are possible to solve analytically (by integration). But when we encounter equations that are hard to solve analytically, we can instead solve them **numerically**. Numerically solving a differential equation means calculating an approximate solution for the variable of interest as it evolves in time, using the information contained in the equation about the *rate of change* of the variable at each time point. 
+
+Let's use the logistic equation as an example again. Like last class, we will plot $dN/dt$ vs. $N$.
+
+```{r}
+library(ggplot2) 
+
+N_seq <-
+  seq(-20, 100) # make a sequence of N values to plot dN/dt against
+  
+logistic_eqn <-  function(N, r, K) {
+  # calculate dN/dt for the logistic equation
+  
+  # r: growth rate (divisions per hour) 
+  # K: carrying capacity
+  # N: population size
+  return(r * N * (1 - N / K))
+  }
+  
+dN_dt <- logistic_eqn(N_seq, r = 0.5, K = 80)
+  
+qplot(N_seq, dN_dt) +
+  geom_line(aes(x = N_seq, y = 0)) # a line at zero for visual aid
+```
+
+$dN/dt$ is the **slope** of $N$ with respect to $t$ at a given value of $N$. 
+We can use the slope as an update rule, exactly like the recursion equation we 
+wrote last class that in the limit became a derivative:
+
+$$lim_{\Delta t \to 0}\frac{N_{t+\Delta t} - N_t}{\Delta t} = \frac{dN}{dt}$$ 
+
+Given a starting point $N_t$, we can *approximate* $N_{t+\Delta t}$ using the differential equation:
+
+$$\frac{N_{t+\Delta t} - N_t}{\Delta t} \approx \frac{dN}{dt} $$
+
+$$N_{t+\Delta t} \approx N_t + \frac{dN}{dt} \Delta t$$
+
+To generate a solution for $N$ as it varies with $t$, we loop through the recursion relation above,
+updating $N$ at each timestep. 
+
+The image below is a cartoon of what this looks like in practice. Starting from a point $A_0$, we use
+the derivative to tell us in what direction we should take our next step. $\Delta t$ is a parameter
+we can choose that determines how large of a step we take. 
+
+This process has several names: **Euler's method** after Leonhard Euler who wrote about it 
+in about 1770, or **forward finite difference method**, referring to the process of stepping 
+forward in time in small (*finite*) increments (*difference*). 
+
+![Euler's method: blue is the true solution, red is the approximate solution. Image source: https://commons.wikimedia.org/w/index.php?curid=2143753](412px-Euler_method.png)
```

Comment:
Instead of uploading the png to the repo, you could just link directly to the website.

```
![Euler's method: blue is the true solution, red is the approximate solution. Image source: Wikimedia](https://upload.wikimedia.org/wikipedia/commons/thumb/1/10/Euler_method.svg/412px-Euler_method.svg.png)
```